src/DynamoCore/Search/SearchDictionary.cs

```diff
@@ -101,8 +101,12 @@ public void Add(V value, IEnumerable<string> tags, double weight = 1)
                 entryDictionary[value] = keys;
                 OnEntryAdded(value);
             }
-            foreach (var tag in tags.Select(x => x.ToLower()))
-                keys[tag] = weight;
+
+            for (int i = 0; i < tags.Count(); i++)
+            {
+                var tag = tags.ElementAt(i).ToLower();
+                keys[tag] = weights.ElementAt(i);
+            }
```

Comment:
`tags.Count()` will be called multiple times, which forces multiple enumeration on `tags` -- we can either cache the count value, or use a `foreach`. Also, this assumes that both `tags` and `weights` are of the same length, we need to handle unequal sized list cases.